### PR TITLE
Fee estimation refactor

### PR DESCRIPTION
**Pull request by gnardini**
### Description

Refactored how fee estimation works and added the fee value to the total amount we show in the confirmation screen.

### Other changes

Uncommented a few `.only` that snuck up in previous PRs 🤦‍♂️ 

### Tested

Manually looking at the logs to see that the estimations are fired at the right time and with unit tests.

### How others should test

Go through the new send flow and see that the tx fees are estimated correctly.

### Related issues

- Part of #1126

### Backwards compatibility

N/A

### CODE REVIEW

File: packages/mobile/src/fees/reducer.ts [hunk 1]
```diff
@@ -1,51 +1,82 @@
...
-export const reducer = combineReducers({ estimates: estimatesReducer })
+const rehydrate = createAction<any>(REHYDRATE)
+export const estimateFee = createAction<EstimateFeeAction>('FEES/ESTIMATE_FEE')
+export const feeEstimated = createAction<{
+  tokenAddress: string
+  feeType: FeeType
+  estimation: FeeEstimateState
+}>('FEES/FEE_ESTIMATED')
+
+export const reducer = createReducer(initialState, (builder) => {
```
Comment (jeanregisser) [inline packages/mobile/src/fees/reducer.ts:47]: Fine for now, but in the future we should favor `createSlice` instead.
See https://github.com/valora-inc/wallet/issues/1442

File: packages/mobile/src/fees/reducer.ts [hunk 2]
```diff
@@ -1,51 +1,82 @@
...
+  builder
+    .addCase(rehydrate, (state, action) => {
+      // hack to allow rehydrate actions here
+      const hydrated = getRehydratePayload((action as unknown) as RehydrateAction, 'fees')
+      return {
+        ...state,
+        ...hydrated,
+      }
+    })
```
Comment (jeanregisser) [inline packages/mobile/src/fees/reducer.ts:56]: This shouldn't be needed. All reducers are rehydrated by default.

File: packages/mobile/src/redux/store.ts [hunk 3]
```diff
@@ -19,10 +19,10 @@ let lastEventTime = Date.now()
 
 const persistConfig: any = {
   key: 'root',
-  version: 20, // default is -1, increment as we make migrations
+  version: 21, // default is -1, increment as we make migrations
   keyPrefix: `reduxStore-`, // the redux-persist default is `persist:` which doesn't work with some file systems.
   storage: FSStorage(),
-  blacklist: ['geth', 'networkInfo', 'alert', 'fees', 'imports'],
```
Comment (jeanregisser) [inline packages/mobile/src/redux/store.ts:25]: Why should we persist fees now?

File: packages/mobile/src/send/SendConfirmation.tsx [hunk 4]
```diff
@@ -263,7 +263,7 @@ function SendConfirmation(props: Props) {
         confirmButton={{
           action: onSendClick,
           text: isInvite ? t('sendAndInvite') : t('send'),
-          disabled: isSending || !!feeError,
```
Comment (jeanregisser) [inline packages/mobile/src/send/SendConfirmation.tsx:266]: Should we really move forward if we couldn't calculate the fee?

File: packages/mobile/src/send/SendConfirmation.tsx [hunk 5]
```diff
@@ -170,17 +166,21 @@ function SendConfirmation(props: Props) {
         <FeeDrawer
           testID={'feeDrawer/SendConfirmation'}
           isEstimate={true}
-          currency={feeInfo?.currency}
+          currency={Currency.Dollar}
```
Comment (jeanregisser) [inline packages/mobile/src/send/SendConfirmation.tsx:169]: Shouldn't this be `feeCurrency`?

File: packages/mobile/src/send/SendConfirmation.tsx [hunk 6]
```diff
@@ -242,7 +242,7 @@ function SendConfirmation(props: Props) {
         usdAmount,
         comment,
         recipient,
-        feeInfo,
+        feeEstimate?.feeInfo,
```
Comment (jeanregisser) [inline packages/mobile/src/send/SendConfirmation.tsx:245]: I think we shouldn't move forward if `feeInfo` is undefined.
Otherwise we could risk sending a transaction where the actual fee hasn't been approved by the user.
or am I missing something?

File: packages/mobile/src/send/SendConfirmation.test.tsx [hunk 7]
```diff
@@ -124,42 +123,47 @@ describe('SendConfirmation', () => {
...
 
-    // NOTE: CELO fees are currently not combined into the total.
-    // TODO: This should equal more than $1.33, depending on the CELO fee value.
+    // Subtotal is $1.33, which is added the fee amount.
     const totalComponent = getByTestId('TotalLineItem/Total')
-    expect(getElementText(totalComponent)).toEqual('$1.33')
+    expect(getElementText(totalComponent)).toEqual('$1.36')
   })
 
-  it('shows a generic `calculateFeeFailed` error when fee estimate fails due to an unknown error', async () => {
```
Comment (jeanregisser) [inline packages/mobile/src/send/SendConfirmation.test.tsx:135]: Do you think showing `---` is enough for users to understand there was an error?

File: packages/mobile/src/send/SendAmountLegacy/SendAmount.test.tsx [hunk 8]
```diff
@@ -144,7 +148,7 @@ describe('SendAmountLegacy', () => {
           dismissAfter: null,
           displayMethod: ErrorDisplayType.BANNER,
           message: i18n.t('needMoreFundsToSend', {
-            amountNeeded: '106.9852',
```
Comment (jeanregisser) [inline packages/mobile/src/send/SendAmountLegacy/SendAmount.test.tsx:147]: Why do we need more now?

File: packages/mobile/src/fees/saga.test.ts [hunk 9]
```diff
@@ -1,79 +1,230 @@
...
+            error: true,
+            usdFee: null,
+            lastUpdated: Date.now(),
+          },
+        })
       )
       .run()
   })
 
-  it("doesn't calculates fee if the balance is zero", async () => {
```
Comment (jeanregisser) [inline packages/mobile/src/fees/saga.test.ts:70]: How does it handle the 0 balance case?

File: packages/mobile/src/fees/saga.ts [hunk 10]
```diff
@@ -23,88 +37,183 @@ export interface FeeInfo {
   currency: Currency
 }
 
-// TODO(victor): This fee caching mechansim is only being used by the balance check on the send
-// amount entry screen. In an effort to standardize and improve fee estimation, we should either
-// update and use this mechanism everywhere or remove it in favor of another solution.
-// Cache of the gas estimates for common tx types
-// Prevents us from having to recreate txs and estimate their gas each time
-const feeGasCache = new Map<FeeType, BigNumber>()
```
Comment (jeanregisser) [inline packages/mobile/src/fees/saga.ts:31]: Are we still caching fees?

**Review by jeanregisser — COMMENTED** [review]
Great work! 🎉 
From what I see it fixes a bunch of edge cases we had. 

And I have a bunch of questions now 🙂

File: packages/mobile/src/components/TokenTotalLineItem.test.tsx [hunk 11]
```diff
@@ -20,11 +20,13 @@ describe('TokenTotalLineItem', () => {
     tokenAddress = defaultTokenAddress,
     localCurrencyCode = LocalCurrencyCode.BRL,
     localCurrencyExchangeRate = '1.5',
+    feeToAddInUsd = undefined,
```
[the remark this answers is not in the document]
Comment (gnardini) [reply packages/mobile/src/components/TokenTotalLineItem.test.tsx:23]: Oh, I missed this, thanks!

File: packages/mobile/src/fees/reducer.ts (hunk 1 above)
Comment (gnardini) [reply packages/mobile/src/fees/reducer.ts:47]: Forgot about this 😓  Will keep in mind

File: packages/mobile/src/redux/store.ts (hunk 3 above)
Comment (gnardini) [reply packages/mobile/src/redux/store.ts:25]: I don't know why we weren't doing it before. The stored fee has a `lastUpdated` field that is used to refresh the fee after some time passed, I don't see why restarting the app should change that. I'm ok adding it back though if you think we shouldn't persist it, no strong opinion on my side.

File: packages/mobile/src/send/SendConfirmation.tsx (hunk 4 above)
Comment (gnardini) [reply packages/mobile/src/send/SendConfirmation.tsx:266]: Well, I changed this thinking of my own personal case, I would prefer to be able to send the tx even if fee estimation fails, but I guess that might not be true to everyone and we shouldn't change it without discussing, I'll revert this

File: packages/mobile/src/send/SendConfirmation.tsx (hunk 6 above)
Comment (gnardini) [reply packages/mobile/src/send/SendConfirmation.tsx:245]: No, you're correct, see my other comment, fixed!

File: packages/mobile/src/send/SendConfirmation.test.tsx (hunk 7 above)
Comment (gnardini) [reply packages/mobile/src/send/SendConfirmation.test.tsx:135]: Not really, no. When fee estimation fails the user also gets an error thing on the top saying it failed, but that happens in the previous screen so it's hard for them to make the connection. I'm just following how it used to work, but I agree we should probably improve on this, I'll create an issue

File: packages/mobile/src/send/SendAmountLegacy/SendAmount.test.tsx (hunk 8 above)
Comment (gnardini) [reply packages/mobile/src/send/SendAmountLegacy/SendAmount.test.tsx:147]: Need more funds because sending now also includes the fee (it didn't before) so the funds they need are a bit higher.

File: packages/mobile/src/send/SendConfirmation.tsx (hunk 5 above)
Comment (gnardini) [reply packages/mobile/src/send/SendConfirmation.tsx:169]: Yes

File: packages/mobile/src/fees/saga.ts (hunk 10 above)
Comment (gnardini) [reply packages/mobile/src/fees/saga.ts:31]: Yes, but in redux. This was I think legacy or something, it was storing fees in this map which is not a pattern we use anywhere else. The comment said `In an effort to standardize and improve fee estimation, we should either
update and use this mechanism everywhere or remove it in favor of another solution.` I chose the second option, removing it in favor of storing fee estimations on Redux

File: packages/mobile/src/fees/saga.test.ts (hunk 9 above)
Comment (gnardini) [reply packages/mobile/src/fees/saga.test.ts:70]: It throws an error. If the balance is exactly zero there is a check that will fail and mark the estimation with `error` and if it's slightly more than zero but less than the required fee the fee estimation fails

File: packages/mobile/src/redux/store.ts (hunk 3 above)
Comment (jeanregisser) [reply packages/mobile/src/redux/store.ts:25]: Yes no strong opinion, but if we decide to persist it we should be able to recover if it's stored as loading and the app is then killed and restarted.

Comment (gnardini) [reply packages/mobile/src/redux/store.ts:25]: Yeah, I think this is not an issue since we always estimate the fee information if there's stale data in the store, even if it's already set to 'loading' so I don't think the UI can get stuck like that.

File: packages/mobile/src/send/SendConfirmation.test.tsx (hunk 7 above)
Comment (gnardini) [reply packages/mobile/src/send/SendConfirmation.test.tsx:135]: Here's the issue: https://app.zenhub.com/workspaces/retention-squad-sprint-board-612f2ef98ef2b800116e6043/issues/valora-inc/wallet/1549

**Review by jeanregisser — APPROVED** [review]

**Comment by ValoraQA** [comment]
Hi @MuckT, We have verified this issue using iOS Test flight build V1.23.0 & updated status in below Gsheet.

**Sheet Link** :
https://docs.google.com/spreadsheets/d/15Z15zu4roT6XFD1dbi9__c3Q6aQZpIiIZka1rKVrCgw/edit#gid=176031153

**Devices** : iPhone 13 mini (15.1.1), iPhone 7(13.3), One Plus 7T(11.0)

CC: @gnardini 
Thanks..!!

**Comment by ValoraQA** [comment]
Hi @MuckT , We have verified this issue using iOS Test flight build V1.23.0 & Android Internal build V: 1.23.0 and updated status in below Gsheet.

**Sheet Link :**
https://docs.google.com/spreadsheets/d/15Z15zu4roT6XFD1dbi9__c3Q6aQZpIiIZka1rKVrCgw/edit#gid=176031153

**Observation :** If we enter amount as 1$ then estimation fees is seen as $0.001 , if we enter amount as 3$ then also the estimation fees is seen as $0.001 so is this calculation fine or there is some formula by which it is calculated 

Devices : iPhone 13mini(15.1.1) , iPhone SE(14.2) , OnePlus 7T(11.0) , Samsung Galaxy M40 (10.0)
CC: @gnardini
Thanks..!!

**Comment by ValoraQA** [comment]
Hi @MuckT , We have verified this issue using Latest Android Internal build V: 1.23.0 and updated status in below Gsheet.

**Sheet Link :**
https://docs.google.com/spreadsheets/d/15Z15zu4roT6XFD1dbi9__c3Q6aQZpIiIZka1rKVrCgw/edit#gid=176031153

**Observation :** If we enter amount as 1$, 3$, 0.01$ & 0.02$ then estimation fees displayed remains same as $0.001 , so is this calculation fine or there is some formula by which it is calculated.

**New Bug:** [Link](https://github.com/valora-inc/wallet/issues/1649)

**Devices :** Samsung Galaxy A8+(8.0), Redmi Note 5 Pro (9.0)
CC: @gnardini
Thanks..!!
